enhancements/machine-config/admin-defined-reboot-policy.md

```diff
@@ -0,0 +1,264 @@
+
+## Proposal
+
+Create a new rebootpolicy.machineconfiguration.openshift.io CRD, allowing specify user-provided rebootpolicies, and view current version cluster defaults, as follows:
+
+```console
+apiVersion: machineconfiguration.openshift.io/v1
+kind: RebootPolicy
+metadata:
+    name: cluster
+spec:
+    userPolicies:
+      - type: file
+        value: /etc/my-conf
+        Actions:
+            - type: drain
+            - type: reload
+              target: myservice
+            - type: reload
+              target: myotherservice
+      - type: kernelArgument
+        value: myarg
+        actions:
+            - type: none
+      - type: service
+        value: myservice
+         actions:
+            - type: restart
+              target: myservice
+    clusterDefaultPolicies:
+      - type: file
+        value: /etc/mco/internal-registry-pull-secret.json
+        actions:
+            - type: none
+      - type: file
+        value: /var/lib/kubelet/config.json
+        actions:
+            - type: none
+      - type: file
+        value: /etc/machine-config-daemon/no-reboot/containers-gpg.pub
+        actions:
+            - type: reload
+              target: crio.service
+      - type: file
+        value: /etc/containers/policy.json
+        actions:
+            - type: reload
+              target: crio.service
+      - type: file
+        value: /etc/containers/registries.conf
+        actions:
+            - type: special
+```
+
+Which, when applied, will define actions for all future changes to these MachineConfig fields based on the diff of the update (rendered MachineConfigs).
+
+For this initial implementation we will support MachineConfig changes to:
+ - Files
+ - Units
+ - kernelArguments
+ - sshKeys
+And actions for:
+ - None
+ - Draining a node
+ - Reloading a service
+ - Daemon-reload and restarting of a service
+ - Reboot (default)
+ - Special (see workflow description)
+
+In the future, we can extend this in two ways:
+Supporting more changes/actions (e.g. user defined script to run after a certain change)
+Supporting diffing between ostree images (for image based updates) and bootc managed systems (for bifrost)
+
+
+### Workflow Description
+
+An admin of an OCP cluster can at any time create/update/delete RebootPolicy objects to reconfigure how they wish updates to be applied. There isn’t any immediate effect so long as the RebootPolicy CR itself is a valid object.
+
+The MachineConfigController’s drain controller will process any update to the object and validate the RebootPolicy for duplication, action validity (but not correctness), etc. If the object is valid, the drain controller will update the object’s status and observedGeneration to indicate success, and if not, update those and degrade the controller, example:
```

Comment:
Why does this need to be done via a controller rather than at admission time using CEL or a validating webhook?

Reply:
I wanted to leave this open for potentially merging multiple rebootpolicies, sort of like MachineConfigs, since I was a bit ambivalent on having just one.

So in that case, users can still provide any configuration of their own, `cluster` would be immutable and generated by default, and the controller will "merge" them in some fashion.

I suppose if we want to just have one object (or as a field in `MachineConfiguration` as mentioned above) having a CEL or validating webhook will be better. I assume those can't be used to cross-validate against other objects?

Reply:
There are options for cross validating against other objects, we can explore that if it is a requirement.

In general, having the reboot policy defined in a single place would likely prevent duplication anyway no? For example, we can use map keys to make sure certain keys within the lists have unique entries